control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+						podExactNames[podName.Name] = true
+					}
+				}
+			}
+		}
+	}
+
+	//TODO(zalimeni) Do we need to check service-enable label here on service/deployments/replicasets?
+	// Should we expect owner is annotated, and if so, is that consistent for non-replicaset workloads?
+	// If so, update to register and deregister conditionally.
+	if err = r.registerService(ctx, resourceClient, service, podPrefixes, podExactNames); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+
+	return ctrl.Result{}, errs
+}
+
+// allAddresses combines all Endpoints subset addresses to a single set. Service registration by this controller
+// operates independent of health, so we combine them here to simplify iteration.
+func allAddresses(addresses corev1.EndpointSubset) map[corev1.EndpointAddress]any {
+	m := make(map[corev1.EndpointAddress]any)
+	for _, readyAddress := range addresses.Addresses {
+		m[readyAddress] = true
+	}
+
+	for _, notReadyAddress := range addresses.NotReadyAddresses {
+		m[notReadyAddress] = true
+	}
+
+	return m
+}
+
+// getOwnerPrefixFromPodName extracts the owner name prefix from a pod name.
+func getOwnerPrefixFromPodName(podName string) string {
+	podNameParts := strings.Split(podName, "-")
+	return strings.Join(podNameParts[:len(podNameParts)-1], "-")
+}
+
+// getOwnerPrefixFromPod returns the common name prefix of the pod, if the pod is a member of a set with a unique name
+// prefix. Currently, this only applies to ReplicaSets.
+//
+// We have to fetch the owner and check its type because pod names cannot be disambiguated from pod owner names due to
+// the `-` delimiter and unique ID parts also being valid name components.
+//
+// If the pod owner does not have a unique name, the empty string is returned.
+func getOwnerPrefixFromPod(pod *corev1.Pod) string {
+	for _, ref := range pod.OwnerReferences {
+		if ref.Kind == "ReplicaSet" {
+			return ref.Name
+		}
+	}
+	return ""
+}
+
+// registerService creates a Consul service registration from the provided Kuberetes service and endpoint information.
+func (r *Controller) registerService(ctx context.Context, resourceClient pbresource.ResourceServiceClient, service corev1.Service, podPrefixes, podExactNames map[string]any) error {
+	serviceResource := r.getServiceResource(
+		&pbcatalog.Service{
+			Workloads:  getWorkloadSelector(podPrefixes, podExactNames),
+			Ports:      getServicePorts(service),
+			VirtualIps: r.getServiceVIPs(service),
+		},
+		service.Name, // Consul and Kubernetes service name will always match
+		r.getConsulNamespace(service.Namespace),
+		r.getConsulPartition(),
+		getServiceMeta(service),
+	)
+
+	r.Log.Info("registering service with Consul", getLogFieldsForResource(serviceResource.Id)...)
+	//TODO(zalimeni) Read from Consul or some other state to determine whether a write is needed.
+	// For now, we blindly write state on each reconcile to simplify implementation.
+	_, err := resourceClient.Write(ctx, &pbresource.WriteRequest{Resource: serviceResource})
```

Comment:
My thinking on this r/n is that we should just try to read the existing Service, and if we find it, check for total equality minus resource version. If there's a diff, rewrite.

If we move to an `EndpointSlices`/`Pods`-based approach in the future, we'll need to do some sort of checking for stale workload selector values, but since `Endpoints` gives us a full snapshot up to the 1K pod limit, we should be able to just upsert today w/o being concerned with drift.

Reply:
Was there a specific case you were trying to address by taking the diff and rewriting it? There was some discussion of running multiple versions of the controller concurrently in the future, and I wonder if this could create flapping with different workers fighting over the same service.

I think I naively assumed that if the write doesn't error, the resource service has persisted the service. Some future work is planned for later releases to sync the status of Consul resources back to Kubernetes resources. Do you think this would address any of your concerns?

Reply:
> Was there a specific case you were trying to address by taking the diff and rewriting it?

Mainly just avoiding needless writes - TBF I'm not fully up to speed on the resource API and state store changes, but my basic understanding is reads are cheaper, and we are theoretically capped in the 2-3K/s for writes. This seemed like an easy way to avoid needless clobbering, esp. when potentially many reconciles loops attempt to write at once w/o versions and the baked-in CAS retries compete w/ each other.

>  There was some discussion of running multiple versions of the controller concurrently in the future, and I wonder if this could create flapping with different workers fighting over the same service.

I think the flapping would occur regardless of pre-reads? But that's a good callout. I'd hope we'd come up w/ some one-way version tagging similar or identical to the `managed-by` tag if needed to avoid issues from that. (Do you have a link to that conversation by chance?)

> I think I naively assumed that if the write doesn't error, the resource service has persisted the service. Some future work is planned for later releases to sync the status of Consul resources back to Kubernetes resources. Do you think this would address any of your concerns?

That's my assumption too 👍🏻 ATM I'm only thinking of debouncing, though bidirectional sync would add complexity to that.

Reply:
I think it was in a casual Zoom, so I don't happen to have a link to the convo.

Reply:
Asked 🧵 [here](https://hashicorp.slack.com/archives/C04AM35BFQT/p1694005306847929) to seek wisdom

Reply:
Summary from Slack: not entirely conclusive, but we agree early optimization should be generally avoided until we get a better idea of write performance in the wild.

I think it's fair to argue that debouncing rapid identical writes falls under low-hanging fruit to reduce needless load, but is leading me to additional questions about how we'll handle state in resources that may be mutated by multiple sources/data flows (not sure how much of this there will be - could resource `Meta` tagging be an example?), since we also can't just blindly rewrite records in that scenario.

Regardless, this is tagged for follow-up outside this PR.